src/bin.js

```diff
@@ -29,13 +29,32 @@ export default function() {
         x1 = xz[1],
         tz = threshold(values, x0, x1);
 
-    // Convert number of thresholds into uniform thresholds,
-    // and nice the default domain accordingly.
+    // Convert number of thresholds into uniform thresholds, and nice the
+    // default domain accordingly.
     if (!Array.isArray(tz)) {
-      tz = +tz;
-      if (domain === extent) [x0, x1] = nice(x0, x1, tz);
-      tz = ticks(x0, x1, tz);
-      if (tz[tz.length - 1] === x1) tz.pop(); // exclusive
+      const max = x1, tn = +tz;
+      if (domain === extent) [x0, x1] = nice(x0, x1, tn);
+      tz = ticks(x0, x1, tn);
+
+      // If the last threshold is coincident with the domain’s upper bound, the
+      // last bin will be zero-width. If the default domain is used, and this
+      // last threshold is coincident with the maximum input value, we can
+      // extend the niced upper bound by one tick to ensure uniform bin widths;
+      // otherwise, we simply remove the last threshold.
+      if (tz[tz.length - 1] === x1) {
+        if (x1 === max && domain === extent) {
+          const step = tickIncrement(x0, x1, tn);
+          if (isFinite(step)) {
```

Comment:
Not clear to me when tickIncrement would not be finite?

I looked up the documentation says “this method is always guaranteed to return an integer”, a comment that seems slightly misleading since:
- d3.tickIncrement(0, 1, 0) returns Infinity
- d3.tickIncrement(0, 0, 0) returns NaN
- d3.tickIncrement(0, 0, 1) return -Infinity

I haven't been able to have it return 0, a case which is tested L50 below, as well as on https://github.com/d3/d3-array/blob/master/src/ticks.js#L15